closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -474,12 +576,272 @@ JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
 
+    if (traceEnabled) {
+        fprintf(stdout, "%s: %s\n", message, openssl_version);
+        fflush(stdout);
+    }
+
+    return ossl_ver;
+}
+
+static void *
+load_crypto_library(jboolean traceEnabled, const char *libName)
+{
+    void *result = NULL;
+    if ((NULL != libName) && ('\0' != *libName)) {
+#if defined(_AIX)
+        int flags = RTLD_NOW;
+        if (NULL != strrchr(libName, '(')) {
+            flags |= RTLD_MEMBER;
+        }
+        result = dlopen(libName, flags);
+#elif defined(_WIN32) /* defined(_AIX) */
+        result = LoadLibrary(libName);
+#else /* defined(_WIN32) */
+        result = dlopen(libName, RTLD_NOW);
+#endif /* defined(_AIX) */
+    }
+    return result;
+}
+
+/* Look for a crypto library in java.home or the system.
+ * NULL is returned when an appropriate crypto library
+ * cannot be found.
+ */
+static void *
+find_crypto_library(jboolean traceEnabled, const char *chomepath)
+{
+    void *result = NULL;
+    void *prevResult = NULL;
+    size_t numOfLibs = 0;
+    size_t i = 0;
+    long tempVersion = 0;
+    long previousVersion = 0;
+    size_t num_of_generic = 0;
+
+#if defined(_AIX)
+    num_of_generic = 4;
+#elif defined(__linux__) /* defined(_AIX) */
+    num_of_generic = 1;
+#endif /* defined(_AIX) */
+
+    /* Library names for OpenSSL 1.1.1, 1.1.0 and symbolic links:
+     * It is important to preserve the order!!!
+     *
+     * Since there is no indication of the version of general
+     * symlink, they have to be loaded first, so as to compare
+     * with other available options.
+     *
+     * The rest of the library are in descending order, which allows
+     * us two things:
+     * - Stop if a general symlink is loaded and we then find a
+     *   specific version that is higher.
+     * - Stop immediately if a specific version is loaded, as
+     *   anything after that will be a lower version.
+     */
+    static const char * const libNames[] = {
+#if defined(_AIX)
+        "libcrypto.a(libcrypto64.so)",      /* general symlink library name from archive file */
+        "libcrypto64.so",                   /* general symlink library name */
+        "libcrypto.a(libcrypto.so)",        /* general symlink library name from archive file */
+        "libcrypto.so",                     /* general symlink library name */
+        "libcrypto.a(libcrypto64.so.3)",    /* 3.x library name from archive file */
+        "libcrypto64.so.3",                 /* 3.x library name */
+        "libcrypto.a(libcrypto.so.3)",      /* 3.x library name from archive file */
+        "libcrypto.so.3",                   /* 3.x library name */
+        "libcrypto.a(libcrypto64.so.1.1)",  /* 1.1.x library name from archive file */
+        "libcrypto.so.1.1",                 /* 1.1.x library name */
+        "libcrypto.so.1.0.0",               /* 1.0.x library name */
+#elif defined(__APPLE__) /* defined(_AIX) */
+        /* "libcrypto.dylib"                Apple no longer supports loading symlink. */
```

Comment:
I think the entry for the symlink should be included as it was before this change.

Reply:
Moved to the end of the list.

Reply:
That's consistent with the old behavior, nor with the comment above (even if it were more than just a comment here).

Reply:
Do you have a reference to support the claim that symbolic links are not supported on macOS?

Reply:
Apple's release notice at https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-release-notes, says

> check for library presence by attempting to dlopen() the path, which will correctly check for the library in the cache

Since we're using `dlopen()`, I don't see a problem. Have you observed otherwise? If listing the name of what is expected to by a symlink doesn't cause a problem, then it should be first to be consistent with other platforms.

Reply:
If you see [this](https://github.com/eclipse-openj9/openj9/issues/17577#issuecomment-1589427344) and [this](https://github.com/eclipse-openj9/openj9/issues/17577#issuecomment-1589428285) comments, you'll see what we did to test and the failure we were getting.

It was producing a warning that was turned to be fatal after `MacOS 11`. I think it's because the `libcrypto.dylib` is not a soft link by default (I've seen hacks where one would delete the existing one and create a soft link manually and it would work), but rather it's expected to produce a failure. It could work if someone has performed those extra steps, but a failure is produced if the default files are in place.